cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/batch/dataset/input/AppWithMapReduceUsingMultipleInputs.java

```diff
@@ -0,0 +1,143 @@
+import org.apache.hadoop.mapreduce.Job;
+import org.apache.hadoop.mapreduce.Mapper;
+import org.apache.hadoop.mapreduce.Reducer;
+import org.apache.hadoop.mapreduce.lib.input.TextInputFormat;
+import org.apache.hadoop.mapreduce.lib.output.TextOutputFormat;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * App used to test whether M/R can read and perform a join across multiple inputs.
+ */
+public class AppWithMapReduceUsingMultipleInputs extends AbstractApplication {
+
+  public static final String PURCHASES = "purchases";
+  public static final String CUSTOMERS = "customers";
+  public static final String OUTPUT_DATASET = "saturatedRecords";
+
+  @Override
+  public void configure() {
+    setName("AppWithMapReduceUsingMultipleInputs");
+    setDescription("Application with MapReduce job using multiple inputs");
+    createDataset(PURCHASES, "fileSet", FileSetProperties.builder()
+      .setInputFormat(TextInputFormat.class)
+      .build());
+    createDataset(CUSTOMERS, "fileSet", FileSetProperties.builder()
+      .setInputFormat(TextInputFormat.class)
+      .build());
+    createDataset(OUTPUT_DATASET, "fileSet", FileSetProperties.builder()
+      .setOutputFormat(TextOutputFormat.class)
+      .setOutputProperty(TextOutputFormat.SEPERATOR, " ")
+      .build());
+    addMapReduce(new ComputeSum());
+  }
+
+  /**
+   * Computes sum of a customer's spending, while also joining on a lookup table, to get more data.
+   */
+  public static final class ComputeSum extends AbstractMapReduce {
+
+    @Override
+    public void beforeSubmit(MapReduceContext context) throws Exception {
+      Map<String, String> inputArgs = new HashMap<>();
+      FileSetArguments.setInputPath(inputArgs, "inputFile");
+
+      context.addInput(Input.of(PURCHASES, inputArgs), FileMapper.class);
+      // since we set a Mapper class on the job itself, omitting the mapper in the addInput call will default to that
+      context.addInput(Input.of(CUSTOMERS, inputArgs));
+
+      Map<String, String> outputArgs = new HashMap<>();
+      FileSetArguments.setOutputPath(outputArgs, "output");
+      context.addOutput(OUTPUT_DATASET, outputArgs);
+
+      Job job = context.getHadoopJob();
+      job.setMapperClass(FileMapper.class);
+      job.setReducerClass(FileReducer.class);
+    }
+  }
+
+  public static class FileMapper extends Mapper<LongWritable, Text, LongWritable, Text>
+    implements ProgramLifecycle<MapReduceTaskContext> {
+
+    private String source;
+
+    @Override
+    public void initialize(MapReduceTaskContext context) throws Exception {
+      System.setProperty("mapper.initialized", "true");
+    }
+
+    @Override
+    public void destroy() {
+      System.setProperty("mapper.destroyed", "true");
+    }
+
+    @Override
+    protected void setup(Context context) throws IOException, InterruptedException {
+      source = ((co.cask.cdap.api.data.batch.TaggedInputSplit) context.getInputSplit()).getName();
```

Comment:
I can expose the name in a `MapReduceTaskContext#getName()`, but I don't see how I can get away with not setting the `TaggedInputSplit` as the `InputSplit` in the hadoop framework. I see no other way of propagating the information of *which* input it is processing to the Mapper. (since each mapper task configuration just gets the same copy. So we need to put the input-specific information in the InputSplit.